packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1053,6 +1058,13 @@ function updateWorkInProgressHook(): Hook {
   return workInProgressHook;
 }
 
+function unstable_useContextWithBailout<T>(
+  context: ReactContext<T>,
+  compare: (T => mixed) | null,
+): T {
+  return readContextAndCompare(context, compare);
```

Comment:
probably should make a null compare function equivalent to just returning the context value. There is logic elsewhere that checks if a compare is passed in but this can lead to maybe confusing code paths if you start with a compare function and then remove it on update or vice versa. Basically if you are using this hook the compare should be required internally and we can make the argument optional for convenience by mapping it to some intuitive default compare function.

That said since this is a compiler target maybe just make the second argument required?

Reply:
Actually now that I am thinking about the fact that the compiler will always return an array from the compare function there really is no identity mapping. If the context value was a class instance how are you planning on reprsenting that? Just an array with the instance in the first slot? I guess that works and you can opt to use multiple indexes if you detect that there is some kind of object destructuring going on?

Still might be worth determining what a null compare argument means semantically or just make it required to avoid the issue

Reply:
This is where the API is designed around testing over direct usage. The goal is to run an A/B performance test. We want some compiler output like

```
const {foo, bar} = useContextWithBailoutTest(MyContext, (c) => [c.foo, c.bar])
```

Then we'll define that hook in the app with experiment check like

```
function useContextWithBailoutTest(context, compare) {
  const inBailoutExperiment = bailoutExperimentCheck();
  return unstable_useContextWithBailout(context, inBailoutExperiment ? compare : null)
}
```

The experiment check will be stable so we wouldn't go between a compare function and null on any update.

We could alternatively pass in some kind of null function

```
return unstable_useContextWithBailout(context, inBailoutExperiment ? compare : () => {})
```

The goal of passing `null` directly was to prevent setting the extra properties on the dependency and running the compare in propagation for the control side of the test. It's likely negligible in real apps but does show up as additional overhead on benchmarks with very fast updates.

Reply:
Makes sense